From bfa145f521906f0d8e231aa3fd26d20d9bac88ed Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Thu, 19 Oct 2023 14:56:35 +0800
Subject: [PATCH] rss/kronos: Add GPT driver for kronos platform

The GPT driver is copied from corstone1000 platform[0] in
Trusted-Firmware-M source code. The driver will be used
to access the MBR/GPT in RSS flash to parse the partition
information of the signed image to fill the flash map
with the entry data, and then the signed images will be
loaded by RSS using the flash map.

[0]:
platform/ext/target/arm/corstone1000/io/*
platform/ext/target/arm/corstone1000/partition/*
platform/ext/target/arm/corstone1000/soft_crc/*
platform/ext/target/arm/corstone1000/platform.c

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../ext/target/arm/rss/kronos/CMakeLists.txt  |  10 +
 .../target/arm/rss/kronos/bl2/boot_hal_bl2.c  |  65 +++
 .../target/arm/rss/kronos/bl2/flash_map_bl2.c |   2 +-
 .../ext/target/arm/rss/kronos/flash_layout.h  |   2 +
 .../ext/target/arm/rss/kronos/io/io_block.c   | 526 ++++++++++++++++++
 .../ext/target/arm/rss/kronos/io/io_block.h   |  40 ++
 .../ext/target/arm/rss/kronos/io/io_defs.h    |  27 +
 .../ext/target/arm/rss/kronos/io/io_driver.h  |  54 ++
 .../ext/target/arm/rss/kronos/io/io_flash.c   | 182 ++++++
 .../ext/target/arm/rss/kronos/io/io_flash.h   |  37 ++
 .../ext/target/arm/rss/kronos/io/io_storage.c | 289 ++++++++++
 .../ext/target/arm/rss/kronos/io/io_storage.h |  92 +++
 .../ext/target/arm/rss/kronos/partition/efi.h |  37 ++
 .../ext/target/arm/rss/kronos/partition/gpt.c |  57 ++
 .../ext/target/arm/rss/kronos/partition/gpt.h |  51 ++
 .../ext/target/arm/rss/kronos/partition/mbr.h |  29 +
 .../arm/rss/kronos/partition/partition.c      | 323 +++++++++++
 .../arm/rss/kronos/partition/partition.h      |  48 ++
 .../target/arm/rss/kronos/partition/uuid.h    |  76 +++
 platform/ext/target/arm/rss/kronos/platform.c |  86 +++
 platform/ext/target/arm/rss/kronos/platform.h |  29 +
 .../target/arm/rss/kronos/soft_crc/soft_crc.c | 119 ++++
 .../target/arm/rss/kronos/soft_crc/soft_crc.h |  17 +
 23 files changed, 2197 insertions(+), 1 deletion(-)
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_block.c
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_block.h
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_defs.h
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_driver.h
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_flash.c
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_flash.h
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_storage.c
 create mode 100644 platform/ext/target/arm/rss/kronos/io/io_storage.h
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/efi.h
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/gpt.c
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/gpt.h
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/mbr.h
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/partition.c
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/partition.h
 create mode 100644 platform/ext/target/arm/rss/kronos/partition/uuid.h
 create mode 100644 platform/ext/target/arm/rss/kronos/platform.c
 create mode 100644 platform/ext/target/arm/rss/kronos/platform.h
 create mode 100644 platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.c
 create mode 100644 platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.h

diff --git a/platform/ext/target/arm/rss/kronos/CMakeLists.txt b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
index d306ea16f..7082ed9fa 100644
--- a/platform/ext/target/arm/rss/kronos/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
@@ -113,6 +113,13 @@ target_sources(platform_bl2
         ni710ae_lib.c
         fainlight_gic_lib.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
+        platform.c
+        io/io_block.c
+        io/io_flash.c
+        io/io_storage.c
+        soft_crc/soft_crc.c
+        partition/partition.c
+        partition/gpt.c
 )
 
 target_compile_definitions(platform_bl2
@@ -137,6 +144,9 @@ target_include_directories(platform_bl2
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
+        io
+        soft_crc
+        partition
 )
 
 target_compile_definitions(platform_bl2
diff --git a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
index e40dce912..08e7ff122 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
@@ -34,14 +34,71 @@
 #include "crypto_hw.h"
 #include "fih.h"
 #endif /* CRYPTO_HW_ACCELERATOR */
+#include "platform.h"
+#include "partition.h"
 
 extern volatile bool scp_doorbell;
 extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
+extern struct flash_area flash_map[];
+extern const int flash_map_entry_num;
 
 static uint8_t *lcp_measurement;
 static struct boot_measurement_metadata *lcp_measurement_metadata;
 
+#define ARRAY_SIZE(arr) (sizeof(arr)/sizeof((arr)[0]))
 static volatile uint8_t chip_id = 0;
+static const char * const image_names [][2] = {
+    {"tfm_primary", "tfm_secondary"},
+    {"tfm_primary", "tfm_secondary"},
+    {"FIP_A", "FIP_B"},
+    {"si_cl2_primary", "si_cl2_secondary"},
+    {"si_cl1_primary", "si_cl1_secondary"},
+    {"si_cl0_primary", "si_cl0_secondary"},
+    {"lcp_primary", "lcp_secondary"},
+    {"scp_primary", "scp_secondary"},
+};
+
+#define BANK_0                  (0)
+
+static bool fill_rss_flash_map_with_data(uint8_t boot_index) {
+    partition_entry_t *entry = NULL;
+    uint8_t i = 0;
+    uint8_t id = 0;
+    uint8_t flash_id = 0;
+    uint8_t image_id = 0;
+
+    if (boot_index >= NR_OF_FW_BANKS) {
+        BOOT_LOG_ERR("%u is an invalid boot_index, 0 <= boot_index < %u",
+                boot_index, NR_OF_FW_BANKS);
+        return false;
+    }
+
+    for (i = 0; i < flash_map_entry_num; i++) {
+        id = flash_map[i].fa_id;
+        for (image_id = 0; image_id < MCUBOOT_IMAGE_NUMBER; image_id++) {
+            if (image_id == RSS_FIRMWARE_AP_BL2_ID) {
+                continue;
+            }
+
+            if (boot_index == BANK_0) {
+                flash_id = FLASH_AREA_IMAGE_PRIMARY(image_id);
+            } else {
+                flash_id = FLASH_AREA_IMAGE_SECONDARY(image_id);
+            }
+
+            if (id == flash_id) {
+                entry = get_partition_entry(image_names[image_id][boot_index]);
+                flash_map[i].fa_off = (uint32_t)entry->start;
+                flash_map[i].fa_size = (uint32_t)entry->length;
+                if (image_id == RSS_FIRMWARE_NON_SECURE_ID) {
+                    flash_map[i].fa_off += FLASH_S_PARTITION_SIZE;
+                }
+            }
+        }
+    }
+
+    return true;
+}
 
 /* Read Chip ID and store it in a global variable */
 static void read_chip_id()
@@ -396,6 +453,14 @@ int32_t boot_platform_init(void)
         return 1;
     }
 
+    plat_io_storage_init();
+    partition_init(PLATFORM_GPT_IMAGE);
+
+    if (!fill_rss_flash_map_with_data(0)) {
+        BOOT_LOG_ERR("Filling flash map with signed images has failed!");
+        return 1;
+    }
+
     return 0;
 }
 
diff --git a/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c
index e06870e68..c3e6b27f8 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c
@@ -23,7 +23,7 @@ const ARM_DRIVER_FLASH *flash_driver[] = {
 };
 const int flash_driver_entry_num = ARRAY_SIZE(flash_driver);
 
-const struct flash_area flash_map[] = {
+struct flash_area flash_map[] = {
     {
         .fa_id = FLASH_AREA_2_ID,
         .fa_device_id = FLASH_DEVICE_ID,
diff --git a/platform/ext/target/arm/rss/kronos/flash_layout.h b/platform/ext/target/arm/rss/kronos/flash_layout.h
index 19ccaa212..20359e93b 100644
--- a/platform/ext/target/arm/rss/kronos/flash_layout.h
+++ b/platform/ext/target/arm/rss/kronos/flash_layout.h
@@ -252,4 +252,6 @@
 /* Smallest flash programmable unit in bytes */
 #define TFM_HAL_PS_PROGRAM_UNIT         (1)
 
+#define NR_OF_FW_BANKS                  (2)
+
 #endif /* __FLASH_LAYOUT_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/io/io_block.c b/platform/ext/target/arm/rss/kronos/io/io_block.c
new file mode 100644
index 000000000..f9bb1c416
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_block.c
@@ -0,0 +1,526 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <errno.h>
+
+#include "io_defs.h"
+#include "io_driver.h"
+#include "io_storage.h"
+#include "io_block.h"
+
+typedef struct {
+    io_block_dev_spec_t *dev_spec;
+    uintptr_t base;
+    uint32_t file_pos;
+    uint32_t size;
+} block_dev_state_t;
+
+#define is_power_of_2(x) (((x) != 0U) && (((x) & ((x)-1U)) == 0U))
+
+io_type_t device_type_block(void);
+
+static int block_open(io_dev_info_t *dev_info, const uintptr_t spec,
+                      io_entity_t *entity);
+static int block_seek(io_entity_t *entity, int mode, size_t offset);
+static int block_read(io_entity_t *entity, uintptr_t buffer, size_t length,
+                      size_t *length_read);
+static int block_write(io_entity_t *entity, const uintptr_t buffer,
+                       size_t length, size_t *length_written);
+static int block_close(io_entity_t *entity);
+static int block_dev_open(const uintptr_t dev_spec, io_dev_info_t **dev_info);
+static int block_dev_close(io_dev_info_t *dev_info);
+
+static const io_dev_connector_t block_dev_connector = {.dev_open =
+                                                           block_dev_open};
+
+static const io_dev_funcs_t block_dev_funcs = {
+    .type = device_type_block,
+    .open = block_open,
+    .seek = block_seek,
+    .size = NULL,
+    .read = block_read,
+    .write = block_write,
+    .close = block_close,
+    .dev_init = NULL,
+    .dev_close = block_dev_close,
+};
+
+static block_dev_state_t state_pool[MAX_IO_BLOCK_DEVICES];
+static io_dev_info_t dev_info_pool[MAX_IO_BLOCK_DEVICES];
+
+/* Track number of allocated block state */
+static unsigned int block_dev_count;
+
+io_type_t device_type_block(void) { return IO_TYPE_BLOCK; }
+
+/* Locate a block state in the pool, specified by address */
+static int find_first_block_state(const io_block_dev_spec_t *dev_spec,
+                                  unsigned int *index_out) {
+    unsigned int index;
+    int result = -ENOENT;
+
+    for (index = 0U; index < MAX_IO_BLOCK_DEVICES; ++index) {
+        /* dev_spec is used as identifier since it's unique */
+        if (state_pool[index].dev_spec == dev_spec) {
+            result = 0;
+            *index_out = index;
+            break;
+        }
+    }
+    return result;
+}
+
+/* Allocate a device info from the pool and return a pointer to it */
+static int allocate_dev_info(io_dev_info_t **dev_info) {
+    int result = -ENOMEM;
+    assert(dev_info != NULL);
+
+    if (block_dev_count < MAX_IO_BLOCK_DEVICES) {
+        unsigned int index = 0;
+        result = find_first_block_state(NULL, &index);
+        assert(result == 0);
+        /* initialize dev_info */
+        dev_info_pool[index].funcs = &block_dev_funcs;
+        dev_info_pool[index].info = (uintptr_t)&state_pool[index];
+        *dev_info = &dev_info_pool[index];
+        ++block_dev_count;
+    }
+
+    return result;
+}
+
+/* Release a device info to the pool */
+static int free_dev_info(io_dev_info_t *dev_info) {
+    int result;
+    unsigned int index = 0;
+    block_dev_state_t *state;
+    assert(dev_info != NULL);
+
+    state = (block_dev_state_t *)dev_info->info;
+    result = find_first_block_state(state->dev_spec, &index);
+    if (result == 0) {
+        /* free if device info is valid */
+        memset(state, 0, sizeof(block_dev_state_t));
+        memset(dev_info, 0, sizeof(io_dev_info_t));
+        --block_dev_count;
+    }
+
+    return result;
+}
+
+static int block_open(io_dev_info_t *dev_info, const uintptr_t spec,
+                      io_entity_t *entity) {
+    block_dev_state_t *cur;
+    io_block_spec_t *region;
+
+    assert((dev_info->info != (uintptr_t)NULL) && (spec != (uintptr_t)NULL) &&
+           (entity->info == (uintptr_t)NULL));
+
+    region = (io_block_spec_t *)spec;
+    cur = (block_dev_state_t *)dev_info->info;
+    assert(((region->offset % cur->dev_spec->block_size) == 0) &&
+           ((region->length % cur->dev_spec->block_size) == 0));
+
+    cur->base = region->offset;
+    cur->size = region->length;
+    cur->file_pos = 0;
+
+    entity->info = (uintptr_t)cur;
+    return 0;
+}
+
+/* parameter offset is relative address at here */
+static int block_seek(io_entity_t *entity, int mode, size_t offset) {
+    block_dev_state_t *cur;
+
+    assert(entity->info != (uintptr_t)NULL);
+
+    cur = (block_dev_state_t *)entity->info;
+
+    assert((offset >= 0) && ((uint32_t)offset < cur->size));
+    switch (mode) {
+        case IO_SEEK_SET:
+            cur->file_pos = (uint32_t)offset;
+            break;
+        case IO_SEEK_CUR:
+            cur->file_pos += (uint32_t)offset;
+            break;
+        default:
+            return -EINVAL;
+    }
+    assert(cur->file_pos < cur->size);
+    return 0;
+}
+
+/*
+ * This function allows the caller to read any number of bytes
+ * from any position. It hides from the caller that the low level
+ * driver only can read aligned blocks of data. For this reason
+ * we need to handle the use case where the first byte to be read is not
+ * aligned to start of the block, the last byte to be read is also not
+ * aligned to the end of a block, and there are zero or more blocks-worth
+ * of data in between.
+ *
+ * In such a case we need to read more bytes than requested (i.e. full
+ * blocks) and strip-out the leading bytes (aka skip) and the trailing
+ * bytes (aka padding). See diagram below
+ *
+ * cur->file_pos ------------
+ *                          |
+ * cur->base                |
+ *  |                       |
+ *  v                       v<----  length   ---->
+ *  --------------------------------------------------------------
+ * |           |         block#1    |        |   block#n          |
+ * |  block#0  |            +       |   ...  |     +              |
+ * |           | <- skip -> +       |        |     + <- padding ->|
+ *  ------------------------+----------------------+--------------
+ *             ^                                                  ^
+ *             |                                                  |
+ *             v    iteration#1                iteration#n        v
+ *              --------------------------------------------------
+ *             |                    |        |                    |
+ *             |<----  request ---->|  ...   |<----- request ---->|
+ *             |                    |        |                    |
+ *              --------------------------------------------------
+ *            /                   /          |                    |
+ *           /                   /           |                    |
+ *          /                   /            |                    |
+ *         /                   /             |                    |
+ *        /                   /              |                    |
+ *       /                   /               |                    |
+ *      /                   /                |                    |
+ *     /                   /                 |                    |
+ *    /                   /                  |                    |
+ *   /                   /                   |                    |
+ *  <---- request ------>                    <------ request  ----->
+ *  ---------------------                    -----------------------
+ *  |        |          |                    |          |           |
+ *  |<-skip->|<-nbytes->|           -------->|<-nbytes->|<-padding->|
+ *  |        |          |           |        |          |           |
+ *  ---------------------           |        -----------------------
+ *  ^        \           \          |        |          |
+ *  |         \           \         |        |          |
+ *  |          \           \        |        |          |
+ *  buf->offset \           \   buf->offset  |          |
+ *               \           \               |          |
+ *                \           \              |          |
+ *                 \           \             |          |
+ *                  \           \            |          |
+ *                   \           \           |          |
+ *                    \           \          |          |
+ *                     \           \         |          |
+ *                      --------------------------------
+ *                      |           |        |         |
+ * buffer-------------->|           | ...    |         |
+ *                      |           |        |         |
+ *                      --------------------------------
+ *                      <-count#1->|                   |
+ *                      <----------  count#n   -------->
+ *                      <----------  length  ---------->
+ *
+ * Additionally, the IO driver has an underlying buffer that is at least
+ * one block-size and may be big enough to allow.
+ */
+static int block_read(io_entity_t *entity, uintptr_t buffer, size_t length,
+                      size_t *length_read) {
+    block_dev_state_t *cur;
+    io_block_spec_t *buf;
+    io_block_ops_t *ops;
+    int lba;
+    size_t block_size, left;
+    size_t nbytes;  /* number of bytes read in one iteration */
+    size_t request; /* number of requested bytes in one iteration */
+    size_t count;   /* number of bytes already read */
+    /*
+     * number of leading bytes from start of the block
+     * to the first byte to be read
+     */
+    size_t skip;
+
+    /*
+     * number of trailing bytes between the last byte
+     * to be read and the end of the block
+     */
+    size_t padding;
+
+    assert(entity->info != (uintptr_t)NULL);
+    cur = (block_dev_state_t *)entity->info;
+    ops = &(cur->dev_spec->ops);
+    buf = &(cur->dev_spec->buffer);
+    block_size = cur->dev_spec->block_size;
+    assert((length <= cur->size) && (length > 0U) && (ops->read != 0));
+
+    /*
+     * We don't know the number of bytes that we are going
+     * to read in every iteration, because it will depend
+     * on the low level driver.
+     */
+    count = 0;
+    for (left = length; left > 0U; left -= nbytes) {
+        /*
+         * We must only request operations aligned to the block
+         * size. Therefore if file_pos is not block-aligned,
+         * we have to request the operation to start at the
+         * previous block boundary and skip the leading bytes. And
+         * similarly, the number of bytes requested must be a
+         * block size multiple
+         */
+        skip = cur->file_pos & (block_size - 1U);
+
+        /*
+         * Calculate the block number containing file_pos
+         * - e.g. block 3.
+         */
+        lba = (cur->file_pos + cur->base) / block_size;
+
+        if ((skip + left) > buf->length) {
+            /*
+             * The underlying read buffer is too small to
+             * read all the required data - limit to just
+             * fill the buffer, and then read again.
+             */
+            request = buf->length;
+        } else {
+            /*
+             * The underlying read buffer is big enough to
+             * read all the required data. Calculate the
+             * number of bytes to read to align with the
+             * block size.
+             */
+            request = skip + left;
+            request = (request + (block_size - 1U)) & ~(block_size - 1U);
+        }
+        request = ops->read(lba, buf->offset, request);
+
+        if (request <= skip) {
+            /*
+             * We couldn't read enough bytes to jump over
+             * the skip bytes, so we should have to read
+             * again the same block, thus generating
+             * the same error.
+             */
+            return -EIO;
+        }
+
+        /*
+         * Need to remove skip and padding bytes,if any, from
+         * the read data when copying to the user buffer.
+         */
+        nbytes = request - skip;
+        padding = (nbytes > left) ? nbytes - left : 0U;
+        nbytes -= padding;
+
+        memcpy((void *)(buffer + count), (void *)(buf->offset + skip), nbytes);
+
+        cur->file_pos += nbytes;
+        count += nbytes;
+    }
+    assert(count == length);
+    *length_read = count;
+
+    return 0;
+}
+
+/*
+ * This function allows the caller to write any number of bytes
+ * from any position. It hides from the caller that the low level
+ * driver only can write aligned blocks of data.
+ * See comments for block_read for more details.
+ */
+static int block_write(io_entity_t *entity, const uintptr_t buffer,
+                       size_t length, size_t *length_written) {
+    block_dev_state_t *cur;
+    io_block_spec_t *buf;
+    io_block_ops_t *ops;
+    int lba;
+    size_t block_size, left;
+    size_t nbytes;  /* number of bytes read in one iteration */
+    size_t request; /* number of requested bytes in one iteration */
+    size_t count;   /* number of bytes already read */
+    /*
+     * number of leading bytes from start of the block
+     * to the first byte to be read
+     */
+    size_t skip;
+
+    /*
+     * number of trailing bytes between the last byte
+     * to be read and the end of the block
+     */
+    size_t padding;
+    assert(entity->info != (uintptr_t)NULL);
+    cur = (block_dev_state_t *)entity->info;
+    ops = &(cur->dev_spec->ops);
+    buf = &(cur->dev_spec->buffer);
+    block_size = cur->dev_spec->block_size;
+    assert((length <= cur->size) && (length > 0U) && (ops->read != 0) &&
+           (ops->write != 0));
+
+    /*
+     * We don't know the number of bytes that we are going
+     * to write in every iteration, because it will depend
+     * on the low level driver.
+     */
+    count = 0;
+    for (left = length; left > 0U; left -= nbytes) {
+        /*
+         * We must only request operations aligned to the block
+         * size. Therefore if file_pos is not block-aligned,
+         * we have to request the operation to start at the
+         * previous block boundary and skip the leading bytes. And
+         * similarly, the number of bytes requested must be a
+         * block size multiple
+         */
+        skip = cur->file_pos & (block_size - 1U);
+
+        /*
+         * Calculate the block number containing file_pos
+         * - e.g. block 3.
+         */
+        lba = (cur->file_pos + cur->base) / block_size;
+
+        if ((skip + left) > buf->length) {
+            /*
+             * The underlying read buffer is too small to
+             * read all the required data - limit to just
+             * fill the buffer, and then read again.
+             */
+            request = buf->length;
+        } else {
+            /*
+             * The underlying read buffer is big enough to
+             * read all the required data. Calculate the
+             * number of bytes to read to align with the
+             * block size.
+             */
+            request = skip + left;
+            request = (request + (block_size - 1U)) & ~(block_size - 1U);
+        }
+
+        /*
+         * The number of bytes that we are going to write
+         * from the user buffer will depend of the size
+         * of the current request.
+         */
+        nbytes = request - skip;
+        padding = (nbytes > left) ? nbytes - left : 0U;
+        nbytes -= padding;
+
+        /*
+         * If we have skip or padding bytes then we have to preserve
+         * some content and it means that we have to read before
+         * writing
+         */
+        if ((skip > 0U) || (padding > 0U)) {
+            request = ops->read(lba, buf->offset, request);
+            /*
+             * The read may return size less than
+             * requested. Round down to the nearest block
+             * boundary
+             */
+            request &= ~(block_size - 1U);
+            if (request <= skip) {
+                /*
+                 * We couldn't read enough bytes to jump over
+                 * the skip bytes, so we should have to read
+                 * again the same block, thus generating
+                 * the same error.
+                 */
+                return -EIO;
+            }
+            nbytes = request - skip;
+            padding = (nbytes > left) ? nbytes - left : 0U;
+            nbytes -= padding;
+        }
+
+        memcpy((void *)(buf->offset + skip), (void *)(buffer + count), nbytes);
+
+        request = ops->write(lba, buf->offset, request);
+        if (request <= skip) return -EIO;
+
+        /*
+         * And the previous write operation may modify the size
+         * of the request, so again, we have to calculate the
+         * number of bytes that we consumed from the user
+         * buffer
+         */
+        nbytes = request - skip;
+        padding = (nbytes > left) ? nbytes - left : 0U;
+        nbytes -= padding;
+
+        cur->file_pos += nbytes;
+        count += nbytes;
+    }
+    assert(count == length);
+    *length_written = count;
+
+    return 0;
+}
+
+static int block_close(io_entity_t *entity) {
+    entity->info = (uintptr_t)NULL;
+    return 0;
+}
+
+static int block_dev_open(const uintptr_t dev_spec, io_dev_info_t **dev_info) {
+    block_dev_state_t *cur;
+    io_block_spec_t *buffer;
+    io_dev_info_t *info;
+    size_t block_size;
+    int result;
+    assert(dev_info != NULL);
+    result = allocate_dev_info(&info);
+    if (result != 0) return -ENOENT;
+
+    cur = (block_dev_state_t *)info->info;
+    /* dev_spec is type of io_block_dev_spec_t. */
+    cur->dev_spec = (io_block_dev_spec_t *)dev_spec;
+    buffer = &(cur->dev_spec->buffer);
+    block_size = cur->dev_spec->block_size;
+
+    assert((block_size > 0U) && (is_power_of_2(block_size) != 0U) &&
+           ((buffer->length % block_size) == 0U));
+
+    *dev_info = info; /* cast away const */
+    (void)block_size;
+    (void)buffer;
+    return 0;
+}
+
+static int block_dev_close(io_dev_info_t *dev_info) {
+    return free_dev_info(dev_info);
+}
+
+/* Exported functions */
+
+/* Register the Block driver with the IO abstraction */
+int register_io_dev_block(const io_dev_connector_t **dev_con) {
+    int result;
+
+    assert(dev_con != NULL);
+
+    /*
+     * Since dev_info isn't really used in io_register_device, always
+     * use the same device info at here instead.
+     */
+    result = io_register_device(&dev_info_pool[0]);
+    if (result == 0) *dev_con = &block_dev_connector;
+    return result;
+}
diff --git a/platform/ext/target/arm/rss/kronos/io/io_block.h b/platform/ext/target/arm/rss/kronos/io/io_block.h
new file mode 100644
index 000000000..0934d048c
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_block.h
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __IO_BLOCK_H__
+#define __IO_BLOCK_H__
+
+#include "io_storage.h"
+
+/* block devices ops */
+typedef struct io_block_ops {
+    size_t (*read)(int lba, uintptr_t buf, size_t size);
+    size_t (*write)(int lba, const uintptr_t buf, size_t size);
+} io_block_ops_t;
+
+typedef struct io_block_dev_spec {
+    io_block_spec_t buffer;
+    io_block_ops_t ops;
+    size_t block_size;
+} io_block_dev_spec_t;
+
+struct io_dev_connector;
+
+int register_io_dev_block(const struct io_dev_connector **dev_con);
+
+#endif /* __IO_BLOCK_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/io/io_defs.h b/platform/ext/target/arm/rss/kronos/io/io_defs.h
new file mode 100644
index 000000000..5db9e58db
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_defs.h
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __IO_DEFS_H__
+#define __IO_DEFS_H__
+
+#define MAX_IO_DEVICES          (2)
+#define MAX_IO_HANDLES          (2)
+#define MAX_IO_BLOCK_DEVICES    (2)
+#define MAX_IO_FLASH_DEVICES    (2)
+
+#endif /* __IO_DEFS_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/io/io_driver.h b/platform/ext/target/arm/rss/kronos/io/io_driver.h
new file mode 100644
index 000000000..5290089e7
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_driver.h
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2014-2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef __IO_DRIVER_H__
+#define __IO_DRIVER_H__
+
+#include <io_storage.h>
+#include <stdint.h>
+
+/* Generic IO entity structure,representing an accessible IO construct on the
+ * device, such as a file */
+typedef struct io_entity {
+    struct io_dev_info *dev_handle;
+    uintptr_t info;
+} io_entity_t;
+
+/* Device info structure, providing device-specific functions and a means of
+ * adding driver-specific state */
+typedef struct io_dev_info {
+    const struct io_dev_funcs *funcs;
+    uintptr_t info;
+} io_dev_info_t;
+
+/* Structure used to create a connection to a type of device */
+typedef struct io_dev_connector {
+    /* dev_open opens a connection to a particular device driver */
+    int (*dev_open)(const uintptr_t dev_spec, io_dev_info_t **dev_info);
+} io_dev_connector_t;
+
+/* Structure to hold device driver function pointers */
+typedef struct io_dev_funcs {
+    io_type_t (*type)(void);
+    int (*open)(io_dev_info_t *dev_info, const uintptr_t spec,
+                io_entity_t *entity);
+    int (*seek)(io_entity_t *entity, int mode, size_t offset);
+    int (*size)(io_entity_t *entity, size_t *length);
+    int (*read)(io_entity_t *entity, uintptr_t buffer, size_t length,
+                size_t *length_read);
+    int (*write)(io_entity_t *entity, const uintptr_t buffer, size_t length,
+                 size_t *length_written);
+    int (*close)(io_entity_t *entity);
+    int (*dev_init)(io_dev_info_t *dev_info, const uintptr_t init_params);
+    int (*dev_close)(io_dev_info_t *dev_info);
+} io_dev_funcs_t;
+
+/* Operations intended to be performed during platform initialisation */
+
+/* Register an IO device */
+int io_register_device(const io_dev_info_t *dev_info);
+
+#endif /* __IO_DRIVER_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/io/io_flash.c b/platform/ext/target/arm/rss/kronos/io/io_flash.c
new file mode 100644
index 000000000..b8c0ff3ab
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_flash.c
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <errno.h>
+
+#include "io_flash.h"
+#include "Driver_Flash.h"
+#include "io_block.h"
+#include "io_defs.h"
+#include "io_driver.h"
+#include "io_storage.h"
+
+#if MAX_IO_FLASH_DEVICES > MAX_IO_BLOCK_DEVICES
+#error \
+    "FLASH devices are BLOCK devices .. MAX_IO_FLASH_DEVICES should be less or equal to MAX_IO_BLOCK_DEVICES"
+#endif
+
+/* Private Prototypes */
+
+static int flash_dev_open(const uintptr_t dev_spec, io_dev_info_t **dev_info);
+static size_t flash_read(int lba, uintptr_t buf, size_t size, size_t flash_id);
+static size_t flash_write(int lba, const uintptr_t buf, size_t size,
+                          size_t flash_id);
+static size_t flash0_read(int lba, uintptr_t buf, size_t size);
+static size_t flash0_write(int lba, uintptr_t buf, size_t size);
+static size_t flash1_read(int lba, uintptr_t buf, size_t size);
+static size_t flash1_write(int lba, uintptr_t buf, size_t size);
+
+/** Private Data **/
+
+/* Flash device data */
+static const io_dev_connector_t flash_dev_connector = {.dev_open =
+                                                           flash_dev_open};
+static size_t flash_dev_count = 0;
+static io_flash_dev_spec_t *flash_dev_specs[MAX_IO_FLASH_DEVICES];
+
+/* Block device data */
+static io_dev_connector_t block_dev_connectors[MAX_IO_FLASH_DEVICES];
+static io_block_dev_spec_t block_dev_spec[MAX_IO_FLASH_DEVICES];
+
+/* Flash devices read/write function pointers */
+static io_block_ops_t flashs_ops[MAX_IO_FLASH_DEVICES] = {
+    [0] = {.read = flash0_read, .write = flash0_write},
+    [1] = {.read = flash1_read, .write = flash1_write},
+};
+
+/* Flash ops functions */
+static size_t flash_read(int lba, uintptr_t buf, size_t size, size_t flash_id) {
+    ARM_DRIVER_FLASH *flash_driver =
+        ((ARM_DRIVER_FLASH *)flash_dev_specs[flash_id]->flash_driver);
+    ARM_FLASH_INFO *info = flash_driver->GetInfo();
+    uint32_t addr = info->sector_size * lba;
+    uint32_t offset = addr - flash_dev_specs[flash_id]->base_addr;
+    size_t rem = info->sector_count * info->sector_size - offset;
+    size_t cnt = size < rem ? size : rem;
+
+    return flash_driver->ReadData(offset, buf, cnt);
+}
+
+static size_t flash_write(int lba, const uintptr_t buf, size_t size,
+                          size_t flash_id) {
+    ARM_DRIVER_FLASH *flash_driver =
+        ((ARM_DRIVER_FLASH *)flash_dev_specs[flash_id]->flash_driver);
+    ARM_FLASH_INFO *info = flash_driver->GetInfo();
+    int32_t rc = 0;
+    uint32_t addr = info->sector_size * lba;
+    uint32_t offset = addr - flash_dev_specs[flash_id]->base_addr;
+    size_t rem = info->sector_count * info->sector_size - offset;
+    size_t cnt = size < rem ? size : rem;
+
+    flash_driver->EraseSector(offset);
+    rc = flash_driver->ProgramData(offset, buf, cnt);
+    return rc;
+}
+
+/* Flash ops functions wrapper for each device */
+
+static size_t flash0_read(int lba, uintptr_t buf, size_t size) {
+    return flash_read(lba, buf, size, 0);
+}
+
+static size_t flash0_write(int lba, uintptr_t buf, size_t size) {
+    return flash_write(lba, buf, size, 0);
+}
+
+static size_t flash1_read(int lba, uintptr_t buf, size_t size) {
+    return flash_read(lba, buf, size, 1);
+}
+
+static size_t flash1_write(int lba, uintptr_t buf, size_t size) {
+    return flash_write(lba, buf, size, 1);
+}
+
+/**
+ *  Helper function to find the index of stored flash_dev_specs or
+ *  return a free slot in case of a new dev_spec
+ */
+static int find_flash_dev_specs(const uintptr_t dev_spec) {
+    /* Search in the saved ones */
+    for (int i = 0; i < flash_dev_count; ++i) {
+        if (flash_dev_specs[i] != NULL &&
+            flash_dev_specs[i]->flash_driver ==
+                ((io_flash_dev_spec_t *)dev_spec)->flash_driver) {
+            return i;
+        }
+    }
+    /* Find the first empty flash_dev_specs to be used */
+    for (int i = 0; i < flash_dev_count; ++i) {
+        if (flash_dev_specs[i] == NULL) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+/**
+ * This function should be called
+ */
+static int flash_dev_open(const uintptr_t dev_spec, io_dev_info_t **dev_info) {
+    ARM_DRIVER_FLASH *flash_driver;
+    assert(dev_info != NULL);
+    assert(dev_spec != NULL);
+
+    size_t index = find_flash_dev_specs(dev_spec);
+
+    /* Check if Flash ops functions are defined for this flash */
+    assert(flashs_ops[index].read && flashs_ops[index].write);
+
+    flash_dev_specs[index] = dev_spec;
+    flash_driver = flash_dev_specs[index]->flash_driver;
+
+    block_dev_spec[index].block_size = flash_driver->GetInfo()->sector_size;
+    block_dev_spec[index].buffer.offset = flash_dev_specs[index]->buffer;
+    block_dev_spec[index].buffer.length = flash_dev_specs[index]->bufferlen;
+    block_dev_spec[index].ops = flashs_ops[index];
+
+    flash_driver->Initialize(NULL);
+
+    block_dev_connectors[index].dev_open(&block_dev_spec[index], dev_info);
+
+    return 0;
+}
+
+/* Exported functions */
+
+/**
+ * Register the flash device.
+ * Internally it register a block device.
+ */
+int register_io_dev_flash(const io_dev_connector_t **dev_con) {
+    int result;
+
+    if (flash_dev_count >= MAX_IO_FLASH_DEVICES) {
+        return -ENOENT;
+    }
+    assert(dev_con != NULL);
+
+    result = register_io_dev_block(dev_con);
+    if (result == 0) {
+        /* Store the block dev connector */
+        block_dev_connectors[flash_dev_count++] = **dev_con;
+        /* Override dev_con with the flash dev connector */
+        *dev_con = &flash_dev_connector;
+    }
+    return result;
+}
diff --git a/platform/ext/target/arm/rss/kronos/io/io_flash.h b/platform/ext/target/arm/rss/kronos/io/io_flash.h
new file mode 100644
index 000000000..8f27fd248
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_flash.h
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __IO_FLASH_H__
+#define __IO_FLASH_H__
+
+#include "io_storage.h"
+
+typedef struct io_flash_dev_spec {
+    uintptr_t buffer;
+    size_t bufferlen;
+    uint32_t base_addr;
+    uintptr_t flash_driver;
+} io_flash_dev_spec_t;
+
+struct io_dev_connector;
+
+/* Register the flash driver with the IO abstraction internally it register a
+ * block device*/
+int register_io_dev_flash(const struct io_dev_connector **dev_con);
+
+#endif /* __IO_FLASH_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/io/io_storage.c b/platform/ext/target/arm/rss/kronos/io/io_storage.c
new file mode 100644
index 000000000..1dde06779
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_storage.c
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#include "io_defs.h"
+#include "io_driver.h"
+
+/* Storage for a fixed maximum number of IO entities, definable by platform */
+static io_entity_t entity_pool[MAX_IO_HANDLES];
+
+/* Simple way of tracking used storage - each entry is NULL or a pointer to an
+ * entity */
+static io_entity_t *entity_map[MAX_IO_HANDLES];
+
+/* Track number of allocated entities */
+static unsigned int entity_count;
+
+/* Array of fixed maximum of registered devices, definable by platform */
+static const io_dev_info_t *devices[MAX_IO_DEVICES];
+
+/* Number of currently registered devices */
+static unsigned int dev_count;
+
+/* Return a boolean value indicating whether a device connector is valid */
+static bool is_valid_dev_connector(const io_dev_connector_t *dev_con) {
+    return (dev_con != NULL) && (dev_con->dev_open != NULL);
+}
+
+/* Return a boolean value indicating whether a device handle is valid */
+static bool is_valid_dev(const uintptr_t dev_handle) {
+    const io_dev_info_t *dev = (io_dev_info_t *)dev_handle;
+
+    return (dev != NULL) && (dev->funcs != NULL) &&
+           (dev->funcs->type != NULL) && (dev->funcs->type() < IO_TYPE_MAX);
+}
+
+/* Return a boolean value indicating whether an IO entity is valid */
+static bool is_valid_entity(const uintptr_t handle) {
+    const io_entity_t *entity = (io_entity_t *)handle;
+
+    return (entity != NULL) && (is_valid_dev((uintptr_t)entity->dev_handle));
+}
+
+/* Return a boolean value indicating whether a seek mode is valid */
+static bool is_valid_seek_mode(io_seek_mode_t mode) {
+    return ((mode != IO_SEEK_INVALID) && (mode < IO_SEEK_MAX));
+}
+
+/* Open a connection to a specific device */
+static int io_storage_dev_open(const io_dev_connector_t *dev_con,
+                               const uintptr_t dev_spec,
+                               io_dev_info_t **dev_info) {
+    assert(dev_info != NULL);
+    assert(is_valid_dev_connector(dev_con));
+
+    return dev_con->dev_open(dev_spec, dev_info);
+}
+
+/* Set a handle to track an entity */
+static void set_handle(uintptr_t *handle, io_entity_t *entity) {
+    assert(handle != NULL);
+    *handle = (uintptr_t)entity;
+}
+
+/* Locate an entity in the pool, specified by address */
+static int find_first_entity(const io_entity_t *entity,
+                             unsigned int *index_out) {
+    int result = -ENOENT;
+    for (unsigned int index = 0; index < MAX_IO_HANDLES; ++index) {
+        if (entity_map[index] == entity) {
+            result = 0;
+            *index_out = index;
+            break;
+        }
+    }
+    return result;
+}
+
+/* Allocate an entity from the pool and return a pointer to it */
+static int allocate_entity(io_entity_t **entity) {
+    int result = -ENOMEM;
+    assert(entity != NULL);
+
+    if (entity_count < MAX_IO_HANDLES) {
+        unsigned int index = 0;
+        result = find_first_entity(NULL, &index);
+        assert(result == 0);
+        *entity = &entity_pool[index];
+        entity_map[index] = &entity_pool[index];
+        ++entity_count;
+    }
+
+    return result;
+}
+
+/* Release an entity back to the pool */
+static int free_entity(const io_entity_t *entity) {
+    int result;
+    unsigned int index = 0;
+    assert(entity != NULL);
+
+    result = find_first_entity(entity, &index);
+    if (result == 0) {
+        entity_map[index] = NULL;
+        --entity_count;
+    }
+
+    return result;
+}
+
+/* Exported API */
+
+/* Register an io device */
+int io_register_device(const io_dev_info_t *dev_info) {
+    int result = -ENOMEM;
+    assert(dev_info != NULL);
+
+    if (dev_count < MAX_IO_DEVICES) {
+        devices[dev_count] = dev_info;
+        dev_count++;
+        result = 0;
+    }
+
+    return result;
+}
+
+/* Open a connection to an IO device */
+int io_dev_open(const io_dev_connector_t *dev_con, const uintptr_t dev_spec,
+                uintptr_t *handle) {
+    assert(handle != NULL);
+    return io_storage_dev_open(dev_con, dev_spec, (io_dev_info_t **)handle);
+}
+
+/* Initialise an IO device explicitly - to permit lazy initialisation or
+ * re-initialisation */
+int io_dev_init(uintptr_t dev_handle, const uintptr_t init_params) {
+    int result = 0;
+    assert(dev_handle != (uintptr_t)NULL);
+    assert(is_valid_dev(dev_handle));
+
+    io_dev_info_t *dev = (io_dev_info_t *)dev_handle;
+
+    /* Absence of registered function implies NOP here */
+    if (dev->funcs->dev_init != NULL) {
+        result = dev->funcs->dev_init(dev, init_params);
+    }
+
+    return result;
+}
+
+/* Close a connection to a device */
+int io_dev_close(uintptr_t dev_handle) {
+    int result = 0;
+    assert(dev_handle != (uintptr_t)NULL);
+    assert(is_valid_dev(dev_handle));
+
+    io_dev_info_t *dev = (io_dev_info_t *)dev_handle;
+
+    /* Absence of registered function implies NOP here */
+    if (dev->funcs->dev_close != NULL) {
+        result = dev->funcs->dev_close(dev);
+    }
+
+    return result;
+}
+
+/* Synchronous operations */
+
+/* Open an IO entity */
+int io_open(uintptr_t dev_handle, const uintptr_t spec, uintptr_t *handle) {
+    int result;
+    assert((spec != (uintptr_t)NULL) && (handle != NULL));
+    assert(is_valid_dev(dev_handle));
+
+    io_dev_info_t *dev = (io_dev_info_t *)dev_handle;
+    io_entity_t *entity;
+
+    result = allocate_entity(&entity);
+
+    if (result == 0) {
+        assert(dev->funcs->open != NULL);
+        result = dev->funcs->open(dev, spec, entity);
+
+        if (result == 0) {
+            entity->dev_handle = dev;
+            set_handle(handle, entity);
+        } else
+            free_entity(entity);
+    }
+    return result;
+}
+
+/* Seek to a specific position in an IO entity */
+int io_seek(uintptr_t handle, io_seek_mode_t mode, int32_t offset) {
+    int result = -ENODEV;
+    assert(is_valid_entity(handle) && is_valid_seek_mode(mode));
+
+    io_entity_t *entity = (io_entity_t *)handle;
+
+    io_dev_info_t *dev = entity->dev_handle;
+
+    if (dev->funcs->seek != NULL)
+        result = dev->funcs->seek(entity, mode, offset);
+
+    return result;
+}
+
+/* Determine the length of an IO entity */
+int io_size(uintptr_t handle, size_t *length) {
+    int result = -ENODEV;
+    assert(is_valid_entity(handle) && (length != NULL));
+
+    io_entity_t *entity = (io_entity_t *)handle;
+
+    io_dev_info_t *dev = entity->dev_handle;
+
+    if (dev->funcs->size != NULL) result = dev->funcs->size(entity, length);
+
+    return result;
+}
+
+/* Read data from an IO entity */
+int io_read(uintptr_t handle, uintptr_t buffer, size_t length,
+            size_t *length_read) {
+    int result = -ENODEV;
+    assert(is_valid_entity(handle));
+
+    io_entity_t *entity = (io_entity_t *)handle;
+
+    io_dev_info_t *dev = entity->dev_handle;
+
+    if (dev->funcs->read != NULL)
+        result = dev->funcs->read(entity, buffer, length, length_read);
+
+    return result;
+}
+
+/* Write data to an IO entity */
+int io_write(uintptr_t handle, const uintptr_t buffer, size_t length,
+             size_t *length_written) {
+    int result = -ENODEV;
+    assert(is_valid_entity(handle));
+
+    io_entity_t *entity = (io_entity_t *)handle;
+
+    io_dev_info_t *dev = entity->dev_handle;
+
+    if (dev->funcs->write != NULL) {
+        result = dev->funcs->write(entity, buffer, length, length_written);
+    }
+
+    return result;
+}
+
+/* Close an IO entity */
+int io_close(uintptr_t handle) {
+    int result = 0;
+    assert(is_valid_entity(handle));
+
+    io_entity_t *entity = (io_entity_t *)handle;
+
+    io_dev_info_t *dev = entity->dev_handle;
+
+    /* Absence of registered function implies NOP here */
+    if (dev->funcs->close != NULL) result = dev->funcs->close(entity);
+
+    /* Ignore improbable free_entity failure */
+    (void)free_entity(entity);
+
+    return result;
+}
diff --git a/platform/ext/target/arm/rss/kronos/io/io_storage.h b/platform/ext/target/arm/rss/kronos/io/io_storage.h
new file mode 100644
index 000000000..4d1b5a81e
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/io/io_storage.h
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2022-2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __IO_STORAGE_H__
+#define __IO_STORAGE_H__
+
+#include <stddef.h>
+#include <stdint.h>
+
+/* Access modes used when accessing data on a device */
+#define IO_MODE_INVALID (0)
+#define IO_MODE_RO (1 << 0)
+#define IO_MODE_RW (1 << 1)
+
+/* Device type which can be used to enable policy decisions about which device
+ * to access */
+typedef enum {
+    IO_TYPE_INVALID,
+    IO_TYPE_SEMIHOSTING,
+    IO_TYPE_MEMMAP,
+    IO_TYPE_DUMMY,
+    IO_TYPE_FIRMWARE_IMAGE_PACKAGE,
+    IO_TYPE_BLOCK,
+    IO_TYPE_MTD,
+    IO_TYPE_MMC,
+    IO_TYPE_STM32IMAGE,
+    IO_TYPE_ENCRYPTED,
+    IO_TYPE_MAX
+} io_type_t;
+
+/* Modes used when seeking data on a supported device */
+typedef enum {
+    IO_SEEK_INVALID,
+    IO_SEEK_SET,
+    IO_SEEK_END,
+    IO_SEEK_CUR,
+    IO_SEEK_MAX
+} io_seek_mode_t;
+
+/* Connector type, providing a means of identifying a device to open */
+struct io_dev_connector;
+
+/* Block specification - used to refer to data on a device supporting
+ * block-like entities */
+typedef struct io_block_spec {
+    size_t offset;
+    size_t length;
+} io_block_spec_t;
+
+
+/* Open a connection to a device */
+int io_dev_open(const struct io_dev_connector *dev_con,
+                const uintptr_t dev_spec, uintptr_t *handle);
+
+/* Initialise a device explicitly - to permit lazy initialisation or
+ * re-initialisation */
+int io_dev_init(uintptr_t dev_handle, const uintptr_t init_params);
+
+/* Close a connection to a device */
+int io_dev_close(uintptr_t dev_handle);
+
+/* Synchronous operations */
+int io_open(uintptr_t dev_handle, const uintptr_t spec, uintptr_t *handle);
+
+int io_seek(uintptr_t handle, io_seek_mode_t mode, int32_t offset);
+
+int io_size(uintptr_t handle, size_t *length);
+
+int io_read(uintptr_t handle, uintptr_t buffer, size_t length,
+            size_t *length_read);
+
+int io_write(uintptr_t handle, const uintptr_t buffer, size_t length,
+             size_t *length_written);
+
+int io_close(uintptr_t handle);
+
+#endif /* __IO_STORAGE_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/partition/efi.h b/platform/ext/target/arm/rss/kronos/partition/efi.h
new file mode 100644
index 000000000..3b7c4f0cb
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/efi.h
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2021,2023 Linaro Limited
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef DRIVERS_PARTITION_EFI_H
+#define DRIVERS_PARTITION_EFI_H
+
+#include <stdint.h>
+#include <string.h>
+
+#include "uuid.h"
+
+#define EFI_NAMELEN 36
+
+static inline int guidcmp(const void *g1, const void *g2) {
+    return memcmp(g1, g2, sizeof(struct efi_guid));
+}
+
+static inline void *guidcpy(void *dst, const void *src) {
+    return memcpy(dst, src, sizeof(struct efi_guid));
+}
+
+#define EFI_GUID(a, b, c, d0, d1, d2, d3, d4, d5, d6, d7)  \
+    {                                                      \
+        (a) & 0xffffffff, (b)&0xffff, (c)&0xffff, {        \
+            (d0), (d1), (d2), (d3), (d4), (d5), (d6), (d7) \
+        }                                                  \
+    }
+
+#define NULL_GUID                                                            \
+    EFI_GUID(0x00000000, 0x0000, 0x0000, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+             0x00, 0x00)
+
+#endif /* DRIVERS_PARTITION_EFI_H */
diff --git a/platform/ext/target/arm/rss/kronos/partition/gpt.c b/platform/ext/target/arm/rss/kronos/partition/gpt.c
new file mode 100644
index 000000000..ef5165384
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/gpt.c
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2016-2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "gpt.h"
+#include "efi.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+
+static int unicode_to_ascii(unsigned short *str_in, unsigned char *str_out) {
+    uint8_t *name;
+    int i;
+
+    assert((str_in != NULL) && (str_out != NULL));
+
+    name = (uint8_t *)str_in;
+
+    assert(name[0] != '\0');
+
+    /* Check whether the unicode string is valid */
+    for (i = 1; i < (EFI_NAMELEN << 1); i += 2) {
+        if (name[i] != '\0') return -EINVAL;
+    }
+    /* Convert the unicode string to ascii string */
+    for (i = 0; i < (EFI_NAMELEN << 1); i += 2) {
+        str_out[i >> 1] = name[i];
+        if (name[i] == '\0') break;
+    }
+    return 0;
+}
+
+int parse_gpt_entry(gpt_entry_t *gpt_entry, partition_entry_t *entry) {
+    int result;
+
+    assert((gpt_entry != NULL) && (entry != NULL));
+
+    if ((gpt_entry->first_lba == 0) && (gpt_entry->last_lba == 0)) {
+        return -EINVAL;
+    }
+
+    memset(entry, 0, sizeof(partition_entry_t));
+    result = unicode_to_ascii(gpt_entry->name, (uint8_t *)entry->name);
+    if (result != 0) {
+        return result;
+    }
+    entry->start = (uint64_t)gpt_entry->first_lba * PLAT_PARTITION_BLOCK_SIZE;
+    entry->length = (uint64_t)(gpt_entry->last_lba - gpt_entry->first_lba + 1) *
+                    PLAT_PARTITION_BLOCK_SIZE;
+    guidcpy(&entry->part_guid, &gpt_entry->unique_uuid);
+    guidcpy(&entry->type_guid, &gpt_entry->type_uuid);
+
+    return 0;
+}
diff --git a/platform/ext/target/arm/rss/kronos/partition/gpt.h b/platform/ext/target/arm/rss/kronos/partition/gpt.h
new file mode 100644
index 000000000..f4353bff2
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/gpt.h
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2016,2023 ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef GPT_H
+#define GPT_H
+
+#include "efi.h"
+#include "partition.h"
+#include "uuid.h"
+
+#define PARTITION_TYPE_GPT 0xee
+#define GPT_HEADER_OFFSET PLAT_PARTITION_BLOCK_SIZE
+#define GPT_ENTRY_OFFSET (GPT_HEADER_OFFSET + PLAT_PARTITION_BLOCK_SIZE)
+
+#define GPT_SIGNATURE "EFI PART"
+
+typedef struct gpt_entry {
+    struct efi_guid type_uuid;
+    struct efi_guid unique_uuid;
+    unsigned long long first_lba;
+    unsigned long long last_lba;
+    unsigned long long attr;
+    unsigned short name[EFI_NAMELEN];
+} gpt_entry_t;
+
+typedef struct gpt_header {
+    unsigned char signature[8];
+    unsigned int revision;
+    unsigned int size;
+    unsigned int header_crc;
+    unsigned int reserved;
+    unsigned long long current_lba;
+    unsigned long long backup_lba;
+    unsigned long long first_lba;
+    unsigned long long last_lba;
+    struct efi_guid disk_uuid;
+    /* starting LBA of array of partition entries */
+    unsigned long long part_lba;
+    /* number of partition entries in array */
+    unsigned int list_num;
+    /* size of a single partition entry (usually 128) */
+    unsigned int part_size;
+    unsigned int part_crc;
+} gpt_header_t;
+
+int parse_gpt_entry(gpt_entry_t *gpt_entry, partition_entry_t *entry);
+
+#endif /* GPT_H */
diff --git a/platform/ext/target/arm/rss/kronos/partition/mbr.h b/platform/ext/target/arm/rss/kronos/partition/mbr.h
new file mode 100644
index 000000000..f1f3cc514
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/mbr.h
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2016,2023 ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef MBR_H
+#define MBR_H
+
+#define MBR_OFFSET 0
+
+#define MBR_PRIMARY_ENTRY_OFFSET 0x1be
+#define MBR_PRIMARY_ENTRY_SIZE 0x10
+#define MBR_PRIMARY_ENTRY_NUMBER 4
+#define MBR_CHS_ADDRESS_LEN 3
+
+#define MBR_SIGNATURE_FIRST 0x55
+#define MBR_SIGNATURE_SECOND 0xAA
+
+typedef struct mbr_entry {
+    unsigned char status;
+    unsigned char first_sector[MBR_CHS_ADDRESS_LEN];
+    unsigned char type;
+    unsigned char last_sector[MBR_CHS_ADDRESS_LEN];
+    unsigned int first_lba;
+    unsigned int sector_nums;
+} mbr_entry_t;
+
+#endif /* MBR_H */
diff --git a/platform/ext/target/arm/rss/kronos/partition/partition.c b/platform/ext/target/arm/rss/kronos/partition/partition.c
new file mode 100644
index 000000000..fe6f229cd
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/partition.c
@@ -0,0 +1,323 @@
+/*
+ * Copyright (c) 2016-2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "efi.h"
+#include "gpt.h"
+#include "mbr.h"
+#include "partition.h"
+#include "bootutil/bootutil_log.h"
+#include "uart_stdout.h"
+#include "io_storage.h"
+#include "platform.h"
+#include "soft_crc.h"
+
+static uint8_t mbr_sector[PLAT_PARTITION_BLOCK_SIZE];
+static partition_entry_list_t list;
+
+#if LOG_LEVEL >= LOG_LEVEL_DEBUG
+static void dump_entries(int num) {
+    char name[EFI_NAMELEN];
+    int i, j, len;
+
+    BOOT_LOG_INF("Partition table with %d entries:", num);
+    for (i = 0; i < num; i++) {
+        len = snprintf(name, EFI_NAMELEN, "%s", list.list[i].name);
+        for (j = 0; j < EFI_NAMELEN - len - 1; j++) {
+            name[len + j] = ' ';
+        }
+        name[EFI_NAMELEN - 1] = '\0';
+        BOOT_LOG_INF("%d: %s %x%x %d", i + 1, name,
+                (uint32_t)(list.list[i].start >> 32),
+                (uint32_t)list.list[i].start,
+                (uint32_t)(list.list[i].start + list.list[i].length - 4));
+    }
+}
+#else
+#define dump_entries(num) ((void)num)
+#endif
+
+/*
+ * Load the first sector that carries MBR header.
+ * The MBR boot signature should be always valid whether it's MBR or GPT.
+ */
+static int load_mbr_header(uintptr_t image_handle, mbr_entry_t *mbr_entry) {
+    size_t bytes_read;
+    uintptr_t offset;
+    int result;
+
+    assert(mbr_entry != NULL);
+    /* MBR partition table is in LBA0. */
+    result = io_seek(image_handle, IO_SEEK_SET, MBR_OFFSET);
+    if (result != 0) {
+        BOOT_LOG_ERR("Failed to seek (%i)\n", result);
+        return result;
+    }
+
+    result = io_read(image_handle, (uintptr_t)&mbr_sector,
+                     PLAT_PARTITION_BLOCK_SIZE, &bytes_read);
+    if (result != 0) {
+        BOOT_LOG_ERR("Failed to read data (%i)\n", result);
+        return result;
+    }
+
+    /* Check MBR boot signature. */
+    if ((mbr_sector[LEGACY_PARTITION_BLOCK_SIZE - 2] != MBR_SIGNATURE_FIRST) ||
+        (mbr_sector[LEGACY_PARTITION_BLOCK_SIZE - 1] != MBR_SIGNATURE_SECOND)) {
+        BOOT_LOG_ERR("MBR signature isn't correct");
+        return -ENOENT;
+    }
+
+    offset = (uintptr_t)&mbr_sector + MBR_PRIMARY_ENTRY_OFFSET;
+    memcpy(mbr_entry, (void *)offset, sizeof(mbr_entry_t));
+    return 0;
+}
+
+/*
+ * Load GPT header and check the GPT signature and header CRC.
+ * If partition numbers could be found, check & update it.
+ */
+static int load_gpt_header(uintptr_t image_handle) {
+    gpt_header_t header;
+    size_t bytes_read;
+    int result;
+    uint32_t header_crc, calc_crc;
+
+    result = io_seek(image_handle, IO_SEEK_SET, GPT_HEADER_OFFSET);
+    if (result != 0) {
+        return result;
+    }
+    result = io_read(image_handle, (uintptr_t)&header, sizeof(gpt_header_t),
+                     &bytes_read);
+    if ((result != 0) || (sizeof(gpt_header_t) != bytes_read)) {
+        return result;
+    }
+    if (memcmp(header.signature, GPT_SIGNATURE, sizeof(header.signature)) !=
+        0) {
+        return -EINVAL;
+    }
+
+    /*
+     * UEFI Spec 2.8 March 2019 Page 119: HeaderCRC32 value is
+     * computed by setting this field to 0, and computing the
+     * 32-bit CRC for HeaderSize bytes.
+     */
+    header_crc = header.header_crc;
+    header.header_crc = 0U;
+
+    calc_crc = crc32((uint8_t *)&header, DEFAULT_GPT_HEADER_SIZE);
+    if (header_crc != calc_crc) {
+        BOOT_LOG_ERR("Invalid GPT Header CRC: Expected 0x%x but got 0x%x.\n",
+                     header_crc, calc_crc);
+        return -EINVAL;
+    }
+
+    header.header_crc = header_crc;
+
+    /* partition numbers can't exceed PLAT_PARTITION_MAX_ENTRIES */
+    list.entry_count = header.list_num;
+    if (list.entry_count > PLAT_PARTITION_MAX_ENTRIES) {
+        list.entry_count = PLAT_PARTITION_MAX_ENTRIES;
+    }
+    return 0;
+}
+
+static int load_mbr_entry(uintptr_t image_handle, mbr_entry_t *mbr_entry,
+                          int part_number) {
+    size_t bytes_read;
+    uintptr_t offset;
+    int result;
+
+    assert(mbr_entry != NULL);
+    /* MBR partition table is in LBA0. */
+    result = io_seek(image_handle, IO_SEEK_SET, MBR_OFFSET);
+    if (result != 0) {
+        BOOT_LOG_ERR("Failed to seek (%i)\n", result);
+        return result;
+    }
+    result = io_read(image_handle, (uintptr_t)&mbr_sector,
+                     PLAT_PARTITION_BLOCK_SIZE, &bytes_read);
+    if (result != 0) {
+        BOOT_LOG_ERR("Failed to read data (%i)\n", result);
+        return result;
+    }
+
+    /* Check MBR boot signature. */
+    if ((mbr_sector[LEGACY_PARTITION_BLOCK_SIZE - 2] != MBR_SIGNATURE_FIRST) ||
+        (mbr_sector[LEGACY_PARTITION_BLOCK_SIZE - 1] != MBR_SIGNATURE_SECOND)) {
+        return -ENOENT;
+    }
+    offset = (uintptr_t)&mbr_sector + MBR_PRIMARY_ENTRY_OFFSET +
+             MBR_PRIMARY_ENTRY_SIZE * part_number;
+    memcpy(mbr_entry, (void *)offset, sizeof(mbr_entry_t));
+
+    return 0;
+}
+
+static int load_mbr_entries(uintptr_t image_handle) {
+    mbr_entry_t mbr_entry;
+    int i;
+
+    list.entry_count = MBR_PRIMARY_ENTRY_NUMBER;
+
+    for (i = 0; i < list.entry_count; i++) {
+        load_mbr_entry(image_handle, &mbr_entry, i);
+        list.list[i].start = mbr_entry.first_lba * 512;
+        list.list[i].length = mbr_entry.sector_nums * 512;
+        list.list[i].name[0] = mbr_entry.type;
+    }
+
+    return 0;
+}
+
+static int load_gpt_entry(uintptr_t image_handle, gpt_entry_t *entry) {
+    size_t bytes_read;
+    int result;
+
+    assert(entry != NULL);
+    result = io_read(image_handle, (uintptr_t)entry, sizeof(gpt_entry_t),
+                     &bytes_read);
+    if (sizeof(gpt_entry_t) != bytes_read) return -EINVAL;
+    return result;
+}
+
+static int verify_partition_gpt(uintptr_t image_handle) {
+    gpt_entry_t entry;
+    int result, i;
+
+    for (i = 0; i < list.entry_count; i++) {
+        result = load_gpt_entry(image_handle, &entry);
+        assert(result == 0);
+        if (result != 0) {
+            break;
+        }
+        result = parse_gpt_entry(&entry, &list.list[i]);
+        if (result != 0) {
+            break;
+        }
+    }
+    if (i == 0) {
+        return -EINVAL;
+    }
+    /*
+     * Only records the valid partition number that is loaded from
+     * partition table.
+     */
+    list.entry_count = i;
+    dump_entries(list.entry_count);
+
+    return 0;
+}
+
+int load_partition_table(unsigned int image_id) {
+    uintptr_t dev_handle, image_handle, image_spec = 0;
+    mbr_entry_t mbr_entry;
+    int result;
+
+    result = plat_get_image_source(image_id, &dev_handle, &image_spec);
+    if (result != 0) {
+        BOOT_LOG_WRN("Failed to obtain reference to image id=%u (%i)\n", image_id,
+                    result);
+        return result;
+    }
+
+    result = io_open(dev_handle, image_spec, &image_handle);
+    if (result != 0) {
+        BOOT_LOG_WRN("Failed to open image id=%u (%i)\n", image_id, result);
+        return result;
+    }
+
+    result = load_mbr_header(image_handle, &mbr_entry);
+    if (result != 0) {
+        BOOT_LOG_WRN("Loading mbr header failed with image id=%u (%i)\n", image_id,
+                    result);
+        return result;
+    }
+    if (mbr_entry.type == PARTITION_TYPE_GPT) {
+        BOOT_LOG_INF("Loading gpt header");
+        result = load_gpt_header(image_handle);
+        assert(result == 0);
+        if (result != 0) {
+            BOOT_LOG_ERR("Failed load gpt header! %i", result);
+            goto load_partition_table_exit;
+        }
+        result = io_seek(image_handle, IO_SEEK_SET, GPT_ENTRY_OFFSET);
+        assert(result == 0);
+        if (result != 0) {
+            BOOT_LOG_ERR("Failed seek gpt header! %i", result);
+            goto load_partition_table_exit;
+        }
+        result = verify_partition_gpt(image_handle);
+        if (result != 0) {
+            BOOT_LOG_ERR("Failed verify gpt partition %i", result);
+            goto load_partition_table_exit;
+        }
+    } else {
+        result = load_mbr_entries(image_handle);
+    }
+
+load_partition_table_exit:
+    io_close(image_handle);
+    return result;
+}
+
+const partition_entry_t *get_partition_entry(const char *name) {
+    int i;
+
+    for (i = 0; i < list.entry_count; i++) {
+        if (strcmp(name, list.list[i].name) == 0) {
+            return &list.list[i];
+        }
+    }
+    return NULL;
+}
+
+const partition_entry_t *get_partition_entry_by_type(const uuid_t *type_uuid) {
+    int i;
+
+    for (i = 0; i < list.entry_count; i++) {
+        if (guidcmp(type_uuid, &list.list[i].type_guid) == 0) {
+            return &list.list[i];
+        }
+    }
+
+    return NULL;
+}
+
+const partition_entry_t *get_partition_replica_by_type(const uuid_t *type_uuid) {
+    int count = 0;
+    int i;
+
+    for (i = 0; i < list.entry_count; i++) {
+        if (guidcmp(type_uuid, &list.list[i].type_guid) == 0) {
+            if (++count == 2)
+                 return &list.list[i];
+        }
+    }
+
+    return NULL;
+}
+
+const partition_entry_t *get_partition_entry_by_uuid(const uuid_t *part_uuid) {
+    int i;
+
+    for (i = 0; i < list.entry_count; i++) {
+        if (guidcmp(part_uuid, &list.list[i].part_guid) == 0) {
+            return &list.list[i];
+        }
+    }
+
+    return NULL;
+}
+
+const partition_entry_list_t *get_partition_entry_list(void) { return &list; }
+
+void partition_init(unsigned int image_id) { load_partition_table(image_id); }
diff --git a/platform/ext/target/arm/rss/kronos/partition/partition.h b/platform/ext/target/arm/rss/kronos/partition/partition.h
new file mode 100644
index 000000000..3bb347a28
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/partition.h
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2016-2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef PARTITION_H
+#define PARTITION_H
+
+#include <stdint.h>
+
+#include "efi.h"
+#include "uuid.h"
+
+#if !PLAT_PARTITION_MAX_ENTRIES
+#define PLAT_PARTITION_MAX_ENTRIES 20
+#endif /* PLAT_PARTITION_MAX_ENTRIES */
+
+#if !PLAT_PARTITION_BLOCK_SIZE
+#define PLAT_PARTITION_BLOCK_SIZE 512
+#endif /* PLAT_PARTITION_BLOCK_SIZE */
+
+#define LEGACY_PARTITION_BLOCK_SIZE 512
+
+#define DEFAULT_GPT_HEADER_SIZE 92
+
+typedef struct partition_entry {
+    uint64_t start;
+    uint64_t length;
+    char name[EFI_NAMELEN];
+    struct efi_guid part_guid;
+    struct efi_guid type_guid;
+} partition_entry_t;
+
+typedef struct partition_entry_list {
+    partition_entry_t list[PLAT_PARTITION_MAX_ENTRIES];
+    int entry_count;
+} partition_entry_list_t;
+
+int load_partition_table(unsigned int image_id);
+const partition_entry_t *get_partition_entry(const char *name);
+const partition_entry_t *get_partition_entry_by_type(const uuid_t *type_guid);
+const partition_entry_t *get_partition_replica_by_type(const uuid_t *type_uuid);
+const partition_entry_t *get_partition_entry_by_uuid(const uuid_t *part_uuid);
+const partition_entry_list_t *get_partition_entry_list(void);
+void partition_init(unsigned int image_id);
+
+#endif /* PARTITION_H */
diff --git a/platform/ext/target/arm/rss/kronos/partition/uuid.h b/platform/ext/target/arm/rss/kronos/partition/uuid.h
new file mode 100644
index 000000000..51d3ceebc
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/partition/uuid.h
@@ -0,0 +1,76 @@
+/*-
+ * Copyright (c) 2002,2023 Marcel Moolenaar
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $FreeBSD$
+ */
+
+/*
+ * Portions copyright (c) 2014-2020, ARM Limited and Contributors.
+ * All rights reserved.
+ */
+
+#ifndef UUID_H
+#define UUID_H
+
+#include <stdint.h>
+
+/* Length of a node address (an IEEE 802 address). */
+#define _UUID_NODE_LEN 6
+
+/* Length of UUID string including dashes. */
+#define _UUID_STR_LEN 36
+
+/*
+ * See also:
+ *      http://www.opengroup.org/dce/info/draft-leach-uuids-guids-01.txt
+ *      http://www.opengroup.org/onlinepubs/009629399/apdxa.htm
+ *
+ * A DCE 1.1 compatible source representation of UUIDs.
+ */
+struct uuid {
+    uint8_t time_low[4];
+    uint8_t time_mid[2];
+    uint8_t time_hi_and_version[2];
+    uint8_t clock_seq_hi_and_reserved;
+    uint8_t clock_seq_low;
+    uint8_t node[_UUID_NODE_LEN];
+};
+
+struct efi_guid {
+    uint32_t time_low;
+    uint16_t time_mid;
+    uint16_t time_hi_and_version;
+    uint8_t clock_seq_and_node[8];
+};
+
+union uuid_helper_t {
+    struct uuid uuid_struct;
+    struct efi_guid efi_guid;
+};
+
+/* XXX namespace pollution? */
+typedef struct uuid uuid_t;
+
+#endif /* UUID_H */
diff --git a/platform/ext/target/arm/rss/kronos/platform.c b/platform/ext/target/arm/rss/kronos/platform.c
new file mode 100644
index 000000000..623811764
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/platform.c
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#include "stdint.h"
+#include "Driver_Flash.h"
+#include "flash_layout.h"
+#include "io_driver.h"
+#include "io_flash.h"
+#include "io_storage.h"
+#include "platform.h"
+#include "host_base_address.h"
+
+typedef struct {
+    uintptr_t dev_handle;
+    uintptr_t image_spec;
+} platform_image_source_t;
+
+extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
+
+#define FLASH_SECTOR_SIZE 0x1000
+
+static io_dev_connector_t *flash_dev_con;
+static uint8_t local_block_flash[FLASH_SECTOR_SIZE];
+static io_flash_dev_spec_t flash_dev_spec = {
+    .buffer = local_block_flash,
+    .bufferlen = FLASH_SECTOR_SIZE,
+    .flash_driver = &FLASH_DEV_NAME,
+};
+static io_block_spec_t flash_spec = {
+    .length = FLASH_TOTAL_SIZE
+};
+
+static platform_image_source_t platform_image_source[] = {
+    [PLATFORM_GPT_IMAGE] = {
+        .dev_handle = NULL,
+        .image_spec = &flash_spec,
+    }
+};
+
+/* Initialize io storage of the platform */
+int32_t plat_io_storage_init(void)
+{
+    int rc = -1;
+    uintptr_t flash_dev_handle = NULL;
+    uintptr_t flash_handle = NULL;
+
+    rc = register_io_dev_flash((const io_dev_connector_t **) &flash_dev_con);
+    if (rc != 0) {
+        return rc;
+    }
+
+    rc = io_dev_open(flash_dev_con, (const uintptr_t)&flash_dev_spec, &flash_dev_handle);
+    if (rc != 0) {
+        return rc;
+    }
+
+    rc = io_open(flash_dev_handle, (const uintptr_t)&flash_spec, &flash_handle);
+    if (rc != 0) {
+        return rc;
+    }
+
+    rc = io_close(flash_handle);
+    if (rc != 0) {
+        return rc;
+    }
+    /* Update the platform image source that uses the flash with dev handles */
+    platform_image_source[PLATFORM_GPT_IMAGE].dev_handle = flash_dev_handle;
+
+    return rc;
+}
+
+/* Return an IO device handle and specification which can be used to access
+ * an image. This has to be implemented for the GPT parser. */
+int32_t plat_get_image_source(unsigned int image_id, uintptr_t *dev_handle,
+                              uintptr_t *image_spec) {
+    if (image_id >= PLATFORM_IMAGE_COUNT) {
+        return -1;
+    }
+    *dev_handle = platform_image_source[image_id].dev_handle;
+    *image_spec = platform_image_source[image_id].image_spec;
+    return 0;
+}
diff --git a/platform/ext/target/arm/rss/kronos/platform.h b/platform/ext/target/arm/rss/kronos/platform.h
new file mode 100644
index 000000000..a88093ed4
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/platform.h
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __PLATFORM_H__
+#define __PLATFORM_H__
+
+typedef enum {
+    PLATFORM_GPT_IMAGE = 0,
+    PLATFORM_IMAGE_COUNT,
+}platform_image_id_t;
+
+#define FWU_METADATA_TYPE_UUID \
+     ((uuid_t){{0xa0, 0x84, 0x7a, 0x8a}, {0x87, 0x83}, {0xf6, 0x40}, 0xab,  0x41, {0xa8, 0xb9, 0xa5, 0xa6, 0x0d, 0x23}})
+#define PRIVATE_METADATA_TYPE_UUID \
+     ((uuid_t){{0xc3, 0x5d, 0xb5, 0xec}, {0xb7, 0x8a}, {0x84, 0x4a}, 0xab,  0x56, {0xeb, 0x0a, 0x99, 0x74, 0xdb, 0x42}})
+
+/* Initialize io storage of the platform */
+int32_t plat_io_storage_init(void);
+
+/* Return an IO device handle and specification which can be used to access
+ * an image. This has to be implemented for the GPT parser. */
+int32_t plat_get_image_source(unsigned int image_id, uintptr_t *dev_handle,
+                              uintptr_t *image_spec);
+
+#endif /*__PLATFORM_H__*/
diff --git a/platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.c b/platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.c
new file mode 100644
index 000000000..20b962d0e
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.c
@@ -0,0 +1,119 @@
+/* Copyright (C) 1986,2023 Gary S. Brown.  You may use this program, or
+   code or tables extracted from it, as desired without restriction.*/
+
+/* First, the polynomial itself and its table of feedback terms.  The  */
+/* polynomial is                                                       */
+/* X^32+X^26+X^23+X^22+X^16+X^12+X^11+X^10+X^8+X^7+X^5+X^4+X^2+X^1+X^0 */
+/* Note that we take it "backwards" and put the highest-order term in  */
+/* the lowest-order bit.  The X^32 term is "implied"; the LSB is the   */
+/* X^31 term, etc.  The X^0 term (usually shown as "+1") results in    */
+/* the MSB being 1.                                                    */
+
+/* Note that the usual hardware shift register implementation, which   */
+/* is what we're using (we're merely optimizing it by doing eight-bit  */
+/* chunks at a time) shifts bits into the lowest-order term.  In our   */
+/* implementation, that means shifting towards the right.  Why do we   */
+/* do it this way?  Because the calculated CRC must be transmitted in  */
+/* order from highest-order term to lowest-order term.  UARTs transmit */
+/* characters in order from LSB to MSB.  By storing the CRC this way,  */
+/* we hand it to the UART in the order low-byte to high-byte; the UART */
+/* sends each low-bit to hight-bit; and the result is transmission bit */
+/* by bit from highest- to lowest-order term without requiring any bit */
+/* shuffling on our part.  Reception works similarly.                  */
+
+/* The feedback terms table consists of 256, 32-bit entries.  Notes:   */
+/*                                                                     */
+/*  1. The table can be generated at runtime if desired; code to do so */
+/*     is shown later.  It might not be obvious, but the feedback      */
+/*     terms simply represent the results of eight shift/xor opera-    */
+/*     tions for all combinations of data and CRC register values.     */
+/*                                                                     */
+/*  2. The CRC accumulation logic is the same for all CRC polynomials, */
+/*     be they sixteen or thirty-two bits wide.  You simply choose the */
+/*     appropriate table.  Alternatively, because the table can be     */
+/*     generated at runtime, you can start by generating the table for */
+/*     the polynomial in question and use exactly the same "updcrc",   */
+/*     if your application needn't simultaneously handle two CRC       */
+/*     polynomials.  (Note, however, that XMODEM is strange.)          */
+/*                                                                     */
+/*  3. For 16-bit CRCs, the table entries need be only 16 bits wide;   */
+/*     of course, 32-bit entries work OK if the high 16 bits are zero. */
+/*                                                                     */
+/*  4. The values must be right-shifted by eight bits by the "updcrc"  */
+/*     logic; the shift must be unsigned (bring in zeroes).  On some   */
+/*     hardware you could probably optimize the shift in assembler by  */
+/*     using byte-swap instructions.                                   */
+
+/**
+ * The code derived from work by Gary S. Brown.
+*/
+
+#include "soft_crc.h"
+
+const static uint32_t crc_32_tab[] = { /* CRC polynomial 0xedb88320 */
+0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
+0xe963a535, 0x9e6495a3, 0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988,
+0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91, 0x1db71064, 0x6ab020f2,
+0xf3b97148, 0x84be41de, 0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7,
+0x136c9856, 0x646ba8c0, 0xfd62f97a, 0x8a65c9ec, 0x14015c4f, 0x63066cd9,
+0xfa0f3d63, 0x8d080df5, 0x3b6e20c8, 0x4c69105e, 0xd56041e4, 0xa2677172,
+0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b, 0x35b5a8fa, 0x42b2986c,
+0xdbbbc9d6, 0xacbcf940, 0x32d86ce3, 0x45df5c75, 0xdcd60dcf, 0xabd13d59,
+0x26d930ac, 0x51de003a, 0xc8d75180, 0xbfd06116, 0x21b4f4b5, 0x56b3c423,
+0xcfba9599, 0xb8bda50f, 0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924,
+0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d, 0x76dc4190, 0x01db7106,
+0x98d220bc, 0xefd5102a, 0x71b18589, 0x06b6b51f, 0x9fbfe4a5, 0xe8b8d433,
+0x7807c9a2, 0x0f00f934, 0x9609a88e, 0xe10e9818, 0x7f6a0dbb, 0x086d3d2d,
+0x91646c97, 0xe6635c01, 0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e,
+0x6c0695ed, 0x1b01a57b, 0x8208f4c1, 0xf50fc457, 0x65b0d9c6, 0x12b7e950,
+0x8bbeb8ea, 0xfcb9887c, 0x62dd1ddf, 0x15da2d49, 0x8cd37cf3, 0xfbd44c65,
+0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2, 0x4adfa541, 0x3dd895d7,
+0xa4d1c46d, 0xd3d6f4fb, 0x4369e96a, 0x346ed9fc, 0xad678846, 0xda60b8d0,
+0x44042d73, 0x33031de5, 0xaa0a4c5f, 0xdd0d7cc9, 0x5005713c, 0x270241aa,
+0xbe0b1010, 0xc90c2086, 0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
+0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4, 0x59b33d17, 0x2eb40d81,
+0xb7bd5c3b, 0xc0ba6cad, 0xedb88320, 0x9abfb3b6, 0x03b6e20c, 0x74b1d29a,
+0xead54739, 0x9dd277af, 0x04db2615, 0x73dc1683, 0xe3630b12, 0x94643b84,
+0x0d6d6a3e, 0x7a6a5aa8, 0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1,
+0xf00f9344, 0x8708a3d2, 0x1e01f268, 0x6906c2fe, 0xf762575d, 0x806567cb,
+0x196c3671, 0x6e6b06e7, 0xfed41b76, 0x89d32be0, 0x10da7a5a, 0x67dd4acc,
+0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5, 0xd6d6a3e8, 0xa1d1937e,
+0x38d8c2c4, 0x4fdff252, 0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
+0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55,
+0x316e8eef, 0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236,
+0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f, 0xc5ba3bbe, 0xb2bd0b28,
+0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7, 0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d,
+0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a, 0x9c0906a9, 0xeb0e363f,
+0x72076785, 0x05005713, 0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38,
+0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242,
+0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
+0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69,
+0x616bffd3, 0x166ccf45, 0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
+0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc,
+0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
+0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693,
+0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
+0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
+};
+
+#define UPDC32(octet,crc) (crc_32_tab[((crc)\
+     ^ ((uint8_t)octet)) & 0xff] ^ ((crc) >> 8))
+
+static inline uint32_t crc32buf(char *buf, size_t len)
+{
+      register uint32_t oldcrc32;
+
+      oldcrc32 = 0xFFFFFFFF;
+
+      for ( ; len; --len, ++buf)
+      {
+            oldcrc32 = UPDC32(*buf, oldcrc32);
+      }
+
+      return ~oldcrc32;
+}
+
+/* Calculate crc32 */
+uint32_t crc32(const void *buf, size_t len) {
+    return crc32buf(buf, len);
+}
diff --git a/platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.h b/platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.h
new file mode 100644
index 000000000..6fd945ee9
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/soft_crc/soft_crc.h
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __SOFT_CRC_H__
+#define __SOFT_CRC_H__
+
+#include <stddef.h>
+#include <stdint.h>
+
+/* Calculate crc32 */
+uint32_t crc32(const void *buf, size_t len);
+
+#endif /* __SOFT_CRC_H__ */
-- 
2.25.1

